### ISSUE TITLE
Multi node example

### ISSUE THREAD

**Issue by hemildesai** [pytorch-labs/monarch#245]
Hi Monarch team, are there plans to add examples for running monarch on multiple nodes/hosts?

**Comment by hemildesai** [comment]
I guess there's a python binding here - https://github.com/pytorch-labs/monarch/pull/170.

But would be good to have a example showcasing how to add different remote servers to the cluster.

**Comment by hemildesai** [comment]
I was able to build the process_allocator binary, looks like that should be enough. Would be good to include the binary in the python installation.
```
bin/release/process_allocator --help
Runs hyperactor's process allocator

Usage: process_allocator [OPTIONS]

Options:
      --port <PORT>        The port to bind to on [::] (all network
                           interfaces on this host). Same as specifying
                           `--addr=[::]:{port}` [default: 26600]
      --addr <ADDR>        The address to bind to in the form:
                           `{transport}!{address}:{port}` (e.g.
                           `tcp!127.0.0.1:26600`). If specified, `--port`
                           argument is ignored
      --program <PROGRAM>  The path to the binary that this process
                           allocator spawns on an `allocate` request
                           [default: monarch_bootstrap]
  -h, --help               Print help
```

**Comment by suo** [comment]
Yes, we are working on a better example here :) @kiukchung is planning to integrate with torchx as an allocator, so that we can allocate actors on using any cluster scheduler that torchx supports (slurm, kubernetes, ray, etc.)

**Comment by kiukchung** [comment]
I've added support for multi-node jobs via the `TorchXRemoteAllocInitializer` (see: https://github.com/pytorch-labs/monarch/blob/main/python/tests/test_allocator.py#L547), but the UX for OSS requires some refinement.

You can try it out as such (make sure to `pip install torchx-nightly`):

```
import os
from monarch.tools.config import Config, UnnamedAppDef
from monarch.tools import commands
from monarch.tools.components import hyperactor
from monarch._src.actor.allocator import (
    RemoteAllocator,
    TorchXRemoteAllocInitializer,
)

mesh_name="trainer"
num_hosts=2
#  a named resource defined in https://github.com/pytorch/torchx/blob/dc70d904c7b7b232c6f5b86f2aa178fc56864662/torchx/specs/named_resources_generic.py#L46
# can define your own as https://docs.pytorch.org/torchx/latest/advanced.html#registering-named-resources
host_type = "gpu.small" 

appdef = hyperactor.proc_mesh( # soon to be renamed to host_mesh in https://github.com/pytorch-labs/monarch/pull/669
   image="/home/$USER", 
   meshes=[f"{mesh_name}:{num_hosts}:{host_type}"],
   env={}, # env vars passed to your program
)

config = Config(
    scheduler="slurm", 
    scheduler_args = {}, # config options listed in https://docs.pytorch.org/torchx/latest/schedulers/slurm.html
    appdef=appdef
)

server_info = commands.get_or_create("test-job-name", config)
    
initializer = TorchXRemoteAllocInitializer(server_info.server_handle)
allocator = RemoteAllocator(
                    world_id="test",
                    initializer=initializer,
                    heartbeat_interval=_100_MILLISECONDS,
)
alloc = await allocator.allocate(AllocSpec(AllocConstraints(), host=1, gpu=4))
proc_mesh = await ProcMesh.from_alloc(alloc)
actor = await proc_mesh.spawn("test_actor", TestActor)
```

If you prefer using k8s, then change the `scheduler="k8s"`  and refer to the instructions in https://docs.pytorch.org/torchx/latest/schedulers/kubernetes.html for setting `scheduler_args`.

**Comment by dz1iang** [comment]
Hello, I have currently launched 4 machines using a Kubernetes (k8s) Training Job. Could you please advise on how to make Monarch detect these 4 machines and schedule tasks onto them? Thank you. @hemildesai @suo @kiukchung

**Comment by hemildesai** [comment]
Thanks @kiukchung for the response. Is there a way to set up a multi node cluster without torchx? i.e. ability to join Monarch controller from already allocated nodes similar to usecase mentioned by @dz1iang

**Comment by dz1iang** [comment]
It seems that I can start the cluster using the bin/release/process_allocator mentioned above and some Python APIs. However, the APIs related to hosts_from_config still appear to be immature for now.

**Comment by colin2328** [comment]
@dz1iang we are improving k8 support, slated to be part of 0.2 release https://github.com/meta-pytorch/monarch/issues/1958

Also see the https://docs.pytorch.org/tutorials/intermediate/monarch_distributed_tutorial.html for an example of a monarchJob on using SLURM. I will link it here https://meta-pytorch.org/monarch/generated/examples/index.html

We will add native k8 example once ready.

@kiukchung  mentioned code path still works.

> ability to join Monarch controller from already allocated nodes similar to usecase mentioned



creating a hostMesh from allocated nodes is supported in the 'bootstrap' API  https://github.com/meta-pytorch/monarch/blob/main/python/monarch/_src/actor/bootstrap.py#L86

**Comment by colin2328** [comment]
Kubernets example is added here: https://meta-pytorch.org/monarch/generated/examples/ddp/kubernetes_ddp.html

Closing this issue, feel free to create a enw one or reopen if you have other questions
